http/src/main/java/org/sapia/ubik/rmi/server/transport/http/JdkClientConnectionPool.java

```diff
@@ -63,23 +104,71 @@ public String getTransportType() {
 
   @Override
   public void release(RmiConnection conn) {
-    pool.release((JdkRmiClientConnection) conn);
+    synchronized (pool) {
+      pool.release((JdkRmiClientConnection) conn);
+      active.remove(conn);
+    }
   }
 
   @Override
   public void invalidate(RmiConnection conn) {
-    pool.invalidate((JdkRmiClientConnection) conn);
+    synchronized (pool) {
+      doInvalidate(conn);
+    }
+  }
+  
+  // --------------------------------------------------------------------------
+  // Visible for testing
+  
+  boolean isActive(RmiConnection conn) {
+    return active.contains(conn);
+  }
+  
+  Pool<JdkRmiClientConnection> getInternalPool() {
+    return pool;
+  }
+  
+  // --------------------------------------------------------------------------
+  // Restricted
+  
+  void terminateTimedOutConnections() {
+    Set<JdkRmiClientConnection> toCheck = new HashSet<>(active.size() + 1);
+    synchronized (pool) {
+      toCheck.addAll(active);
+    }
+    toCheck.forEach(c -> {
+      if (c.isInReadTimeout()) {
+        log.warning("Invalidating connection to %s since it is deemed in a read timeout situation", c.getServerAddress());
```

Comment:
to be on the safe side, I would put a try catch here. Since this is out of the synchronized block, there is a potential case that the connection to check might not be in the pool anymore (not likely though, but possible)

Reply:
Fixed that